beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -0,0 +1,400 @@
+	randGenerator.Shuffle(len(peers), func(i, j int) {
+		peers[i], peers[j] = peers[j], peers[i]
+	})
+
+	p2pRequests := new(sync.WaitGroup)
+	errChan := make(chan error)
+	blocksChan := make(chan []*eth.SignedBeaconBlock)
+
+	p2pRequests.Add(len(peers))
+	go func() {
+		p2pRequests.Wait()
+		close(blocksChan)
+	}()
+
+	// Short circuit start far exceeding the highest finalized epoch in some infinite loop.
+	highestFinalizedSlot := helpers.StartSlot(finalizedEpoch + 1)
+	if start > highestFinalizedSlot {
+		return nil, errors.Errorf("did not expect request for start slot of %d which is greater than expected %d", start, highestFinalizedSlot)
+	}
+
+	// Spread load evenly among available peers.
+	perPeerCount := count / uint64(len(peers))
+	remainder := int(count % uint64(len(peers)))
+	log.WithFields(logrus.Fields{
+		"start":        start,
+		"count":        count,
+		"perPeerCount": perPeerCount,
+		"remainder":    remainder,
+	}).Debug("Distribute request among available peers")
+	for i, pid := range peers {
+		start, step := start+uint64(i)*step, step*uint64(len(peers))
+
+		// If the count was divided by an odd number of peers, there will be some blocks
+		// missing from the first requests so we accommodate that scenario.
+		count := perPeerCount
+		if i < remainder {
+			count++
+		}
+		// Asking for no blocks may cause the client to hang. This should never happen and
+		// the peer may return an error anyway, but we'll ask for at least one block.
+		if count == 0 {
+			count++
+		}
+
+		go func(ctx context.Context, pid peer.ID) {
+			defer p2pRequests.Done()
+
+			blocks, err := f.requestBeaconBlocksByRange(ctx, pid, root, start, step, count)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			blocksChan <- blocks
+		}(ctx, pid)
+	}
+
+	var unionRespBlocks []*eth.SignedBeaconBlock
+	for {
+		select {
+		case err := <-errChan:
+			return nil, err
+		case resp, ok := <-blocksChan:
+			if ok {
+				unionRespBlocks = append(unionRespBlocks, resp...)
+			} else {
+				sort.Slice(unionRespBlocks, func(i, j int) bool {
+					return unionRespBlocks[i].Block.Slot < unionRespBlocks[j].Block.Slot
+				})
+				return unionRespBlocks, nil
+			}
+		}
+	}
+}
+
+// requestBeaconBlocksByRange prepares BeaconBlocksByRange request, and handles possible stale peers (by resending the request).
+func (f *blocksFetcher) requestBeaconBlocksByRange(ctx context.Context, pid peer.ID, root []byte, start, step, count uint64) ([]*eth.SignedBeaconBlock, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	req := &p2ppb.BeaconBlocksByRangeRequest{
+		HeadBlockRoot: root,
+		StartSlot:     start,
+		Count:         count,
+		Step:          step,
+	}
+
+	resp, respErr := f.requestBlocks(ctx, req, pid)
+	if respErr != nil {
+		// Fail over to some other, randomly selected, peer.
+		root1, _, peers := f.p2p.Peers().BestFinalized(params.BeaconConfig().MaxPeersToSync, helpers.SlotToEpoch(f.headFetcher.HeadSlot()))
```

Comment:
Line very long, can we do headEpoch := helpers.SlotToEpoch(f.headFetcher.HeadSlot()) ?